### PR TITLE
simulation testing

### PR DESCRIPTION
**Pull request by gjermundgaraba**

### CODE REVIEW

**Review by ba1ciu — APPROVED** [review]
Overall I don't see anything that would make me suggest keeping it from merging. There's so much of the boilerplate needed in the project that could be just imported from SDK, that makes me feel anxious, but nothing we can do about it, I guess :) Shared some general remarks, questions and ideas in the comments.

Also, maybe we could use `SimulationState.AppParams` to store seeds/counters that will help finding existing credits easier? `json.RawMessage` is just a `[]byte`, so potentially we could use it to store all generated denoms/abbrevs/serials in there and pick randomly from it. Not very clean, I know, and not sure if possible, but sounds better than adding logic to the keeper if we don't need it for stuff other than simulation?

File: chain/x/plasticcredit/simulation/proposals_test.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/stretchr/testify/require"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+
+	"github.com/EmpowerPlastic/empowerchain/x/plasticcredit"
+	"github.com/EmpowerPlastic/empowerchain/x/plasticcredit/simulation"
+)
+
+func TestProposalMsgs(t *testing.T) {
+	// initialize parameters
+	s := rand.NewSource(1)
```
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/x/plasticcredit/simulation/proposals_test.go:19]: Honestly, I just followed what was done in the SDK repo, but probably just to keep the tests deterministic? Keeps things simple and makes sure you know what will get generated

File: chain/x/plasticcredit/simulation/decoder.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/types/kv"
+
+	"github.com/EmpowerPlastic/empowerchain/x/plasticcredit"
+)
+
+// NewDecodeStore returns a decoder function closure that unmarshal the KVPair's
+// value to the corresponding plasticcredit type.
+func NewDecodeStore(cdc codec.Codec) func(kvA, kvB kv.Pair) string {
+	return func(kvA, kvB kv.Pair) string {
```
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/x/plasticcredit/simulation/decoder.go:16]: No, I don't know at all. It's documented somewhat here, but I still don't understand it: https://docs.cosmos.network/main/building-modules/simulator#store-decoders

File: chain/app/export.go [hunk 3]
```diff
@@ -74,20 +74,21 @@ func (app *EmpowerApp) prepForZeroHeightGenesis(ctx sdk.Context, jailAllowedAddr
 
 	// withdraw all validator commission
 	app.StakingKeeper.IterateValidators(ctx, func(_ int64, val stakingtypes.ValidatorI) (stop bool) {
-		_, err := app.DistrKeeper.WithdrawValidatorCommission(ctx, val.GetOperator())
-		if err != nil {
-			panic(err)
-		}
+		_, _ = app.DistrKeeper.WithdrawValidatorCommission(ctx, val.GetOperator())
```
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/app/export.go:77]: Yeah, this was recently changed to work like this in the simapp in the sdk repo. This caused an error so I just copy-pasted :D

File: chain/app/simulation/app_import_export_test.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		{empowerApp.GetStoreKey(slashingtypes.StoreKey), newApp.GetStoreKey(slashingtypes.StoreKey), [][]byte{}},
+		{empowerApp.GetStoreKey(minttypes.StoreKey), newApp.GetStoreKey(minttypes.StoreKey), [][]byte{}},
+		{empowerApp.GetStoreKey(distrtypes.StoreKey), newApp.GetStoreKey(distrtypes.StoreKey), [][]byte{}},
+		{empowerApp.GetStoreKey(banktypes.StoreKey), newApp.GetStoreKey(banktypes.StoreKey), [][]byte{banktypes.BalancesPrefix}},
+		// {empowerApp.GetStoreKey(paramtypes.StoreKey), newApp.GetStoreKey(paramtypes.StoreKey), [][]byte{}}, // TODO: FIXMEH
+		{empowerApp.GetStoreKey(govtypes.StoreKey), newApp.GetStoreKey(govtypes.StoreKey), [][]byte{}},
+		{empowerApp.GetStoreKey(evidencetypes.StoreKey), newApp.GetStoreKey(evidencetypes.StoreKey), [][]byte{}},
+		{empowerApp.GetStoreKey(capabilitytypes.StoreKey), newApp.GetStoreKey(capabilitytypes.StoreKey), [][]byte{}},
+		{empowerApp.GetStoreKey(authzkeeper.StoreKey), newApp.GetStoreKey(authzkeeper.StoreKey), [][]byte{authzkeeper.GrantKey, authzkeeper.GrantQueuePrefix}},
+		// TODO: ADD EMPOWER MODULES
```
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/app/simulation/app_import_export_test.go:131]: Oppsies, will look at this!

File: chain/x/plasticcredit/simulation/genesis.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		var usedAccounts []simtypes.Account
+		for i := 0; i < numberOfAccounts; i++ {
+			acc := getUnusedAccount(r, accounts, usedAccounts)
+			usedAccounts = append(usedAccounts, acc)
+			var active, retired uint64
+			// On the last account, give the rest of the credits
+			if i == numberOfAccounts-1 {
+				active = activeLeft
+				retired = retiredLeft
+			} else {
```
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/x/plasticcredit/simulation/genesis.go:263]: Good idea. Adding this!

File: chain/x/plasticcredit/simulation/genesis.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+					break
+				}
+			}
+
+			creditCollections = append(creditCollections, plasticcredit.CreditCollection{
+				Denom:     denom,
+				ProjectId: projects[i].Id,
+				TotalAmount: plasticcredit.CreditAmount{
+					Active:  uint64(simtypes.RandIntBetween(r, 0, 1000000)),
+					Retired: uint64(simtypes.RandIntBetween(r, 0, 1000000)),
```
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/x/plasticcredit/simulation/genesis.go:238]: Good idea. Adding this!

**Pull request merged**